### ISSUE TITLE
Are mini "improvements" appreciated?

### ISSUE THREAD

**Issue by rayfordshire** [neomutt/neomutt#3707]
While trying to figure out a S/MIME problem, I'm skimming NeoMutt's source code and sometimes some things pop up, which I think is bad code (examples below) and could easily be fixed by a small diff. Since they do not add something to the code base they might be considered as noise and make it harder to port patches from upstream Mutt.  Therefore, I'd like to know if these kinds of PRs would be welcome?

## Examples:

1)
`gpgme_key_unref(key2);` is called in both branches, therefore the `else` branch is superfluous:
https://github.com/neomutt/neomutt/blob/a92c1521d7d39f085291ba5a276915799a1a3459/ncrypt/crypt_gpgme.c#L670-L681

2)
here `a` and `t` are repurposed multiple times making the code hard to follow.  Introducing new variables/names for the parts would IMHO make it easier to read.
https://github.com/neomutt/neomutt/blob/a92c1521d7d39f085291ba5a276915799a1a3459/ncrypt/crypt_gpgme.c#L984-L987

3)
Restrict scope of `tmp_pbody` to this branch, i.e. not everything is declared at the top like in C89.
https://github.com/neomutt/neomutt/blob/a92c1521d7d39f085291ba5a276915799a1a3459/ncrypt/crypt.c#L340-L355

4)
boolean blindness everywhere
<no example necessary>

**Comment by gahr** [comment]
First off, thanks! I appreciate your attention to the details!

Improvements are appreciated by definition, I guess the question is more whether there’s consensus whether a proposed change *is* an improvement :)

Different people will have different opinions, but a few things I like:
- making variables const, so you know they aren’t re-assigned a few dozens lines down
- using expressions instead of statements: `const char *out = foo ? bar : baz;` vs `const chat *out = NULL; if (foo) out = bar; else out = baz;`
- established conventions: `e` is a `struct Email *`, `m` is a `struct Mailbox *`

and a few things I don’t like:
- comments that repeat in English what the code says in C, or that explain how C works
- unneeded parentheses: I prefer `if (foo(bar) > c - d && baz == bar)` over `if ((foo(bar) > (c - d)) && (baz == bar))` but flatcap won this battle a long time ago :)

IIRC, you said somewhere that you aren’t much experienced in C. I think my advice is to try not to assume that code is bad on the basis that it takes you a while to understand, same as you wouldn’t consider a book in a foreign language “bad” because you have difficulties reading it. Also, C has decades of established untold practices and conventions that you might not be aware of.
I would suggest you start with the obvious things (code duplication, variable reuse, …) and try to avoid the most subjective matters.

Happy to discuss any of this in more detail, perhaps on IRC :)

**Comment by gahr** [comment]
To your points in particular:

1. yes, the else branch can be removed
2. this is a pretty common idiom: when I see it, I recognize it :)  I am not sure I would change this very much, except perhaps for renaming the variables. So what's going on is this: you have a container (here: [Body](https://github.com/neomutt/neomutt/blob/a92c1521d7d39f085291ba5a276915799a1a3459/email/body.h#L71-L72)) that contains elements of the same type of itself. At a certain point you want to wrap a main top-level Body `a` into a new top-level Body (here, the multipart/signed part). So what you do is
    1. build a new top-level: `t = mutt_body_new()`
    2. add the main top-level as a part of the new top-level: `t->parts = a`
    3. now the main thing is the new top-level: `a = t`
3. yes please - "small scopes" is another thing I like very much which I forgot to mention in my list :)

**Comment by flatcap** [comment]
> Are mini "improvements" appreciated?

Yes, absolutely.

> could easily be fixed by a small diff
> might be considered as noise

Early in NeoMutt's history, we tried to keep the changes to a minimum.

Since we gave up on that, we've been aggressively modernising and tidying the code.
(there's still SO much more to do, though :-)

> and make it harder to port patches from upstream Mutt

hehe, that ship has sailed :-)

**Comment by flatcap** [comment]
I've created a branch `devel/trivial`.

@rayfordshire 
There's an invitation for you to join the NeoMutt project.
If you accept, you can push small changes that branch.

@whitney-cumber You're welcome to push to it too.

Then I can review it every few weeks to save having to create PRs for everything.

If you have any questions, gahr and I are on IRC during European working hours.
IRC: `#neomutt` on `irc.libera.chat` ([web client](https://kiwiirc.com/nextclient/irc.libera.chat/#neomutt))

**Comment by flatcap** [comment]
> established conventions: `e` is a `struct Email *`, `m` is a `struct Mailbox *`

Yeah, I keep meaning to document this.
There's:
- Style Guide: https://neomutt.org/dev/code
- Unfinished WIP notes: https://github.com/neomutt/flatcap-brain-dump/blob/main/code/style-guide.md

I'll make some time to put all the ideas in a Discussion.

**Comment by rayfordshire** [comment]
The question has been answered, closing.
